rcl/test/rcl/test_publisher_wait_all_ack.cpp

```diff
@@ -0,0 +1,268 @@
+  constexpr char topic_name[] = "test_wait_for_all_acked";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  publisher_options.qos.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+  publisher_options.qos.depth = 10000;
+  ret = rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  rcl_subscription_options_t subscription_options = rcl_subscription_get_default_options();
+  subscription_options.qos.depth = 1;
+  subscription_options.qos.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+
+#define INIT_SUBSCRIPTION(idx) \
+  rcl_subscription_t subscription ## idx = rcl_get_zero_initialized_subscription(); \
+  ret = rcl_subscription_init( \
+    &subscription ## idx, \
+    this->node_ptr, \
+    ts, \
+    topic_name, \
+    &subscription_options); \
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str; \
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT( \
+  { \
+    ret = rcl_subscription_fini(&subscription ## idx, this->node_ptr); \
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str; \
+  });
+
+  INIT_SUBSCRIPTION(1)
+  INIT_SUBSCRIPTION(2)
+  INIT_SUBSCRIPTION(3)
+
+  ASSERT_TRUE(wait_for_established_subscription(&publisher, 10, 100));
+
+  // Prepare 1MB message
+#define ONE_MEGABYTE (1024 * 1024)
+  char test_string[ONE_MEGABYTE];
+  memset(test_string, 'a', ONE_MEGABYTE);
+  test_string[ONE_MEGABYTE - 1] = '\0';
+  test_msgs__msg__Strings msg;
+  test_msgs__msg__Strings__init(&msg);
+  ASSERT_TRUE(rosidl_runtime_c__String__assign(&msg.string_value, test_string));
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    test_msgs__msg__Strings__fini(&msg);
+  });
+
+  ret = rcl_publish(&publisher, &msg, nullptr);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription1, context_ptr, 10, 100));
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription2, context_ptr, 10, 100));
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription3, context_ptr, 10, 100));
+
+  int i = 0;
+  for (; i < 500; i++) {
+    ret = rcl_publish(&publisher, &msg, nullptr);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+
+  ret = rcl_publisher_wait_for_all_acked(
+    &publisher,
+    RCL_MS_TO_NS(500));
+
+  if (is_cyclonedds) {
+    // cyclonedds use sync publish, so above scenario cannot lead to timeout.
+    EXPECT_EQ(RCL_RET_OK, ret);
+  } else {
+    EXPECT_EQ(RCL_RET_TIMEOUT, ret);
+  }
+  ret = rcl_publisher_wait_for_all_acked(&publisher, -1);
+  EXPECT_EQ(RCL_RET_OK, ret);
+}
+
+TEST_F(
+  CLASSNAME(TestPublisherFixtureSpecial, RMW_IMPLEMENTATION),
+  test_wait_for_all_acked_with_best_effort)
```

Comment:
the test seems to be too complex to only test that calling `rcl_publisher_wait_for_all_acked()` in a best effort publisher does nothing.

I would just create a publisher, no subscription, don't publish anything, and call `rcl_publisher_wait_for_all_acked()` expecting it to return OK.

Reply:
> I would just create a publisher, no subscription, don't publish anything, and call rcl_publisher_wait_for_all_acked() expecting it to return OK.

Even if a reliable publisher is used, the above steps also can return OK.   
So I have to make a complex case to test it.

Reply:
> Even if a reliable publisher is used, the above steps also can return OK.

That could also happen in the current test case if a reliable publisher would be able to send all those messages in 500ms.
The test doesn't seem to be adding much value really.

Reply:
> That could also happen in the current test case if a reliable publisher would be able to send all those messages in 500ms.  

Yes. It is possible.  
But this scenario also leads to the failure of  test case `test_wait_for_all_acked`.  
```
  ret = rcl_publisher_wait_for_all_acked(
    &publisher,
    RCL_MS_TO_NS(500));

  if (is_cyclonedds) {
    // cyclonedds use sync publish, so above scenario cannot lead to timeout.
    EXPECT_EQ(RCL_RET_OK, ret);
  } else {
    EXPECT_EQ(RCL_RET_TIMEOUT, ret);  <== We expect to get RCL_RET_TIMEOUT
  }
```

Anyway, I will simplify this test case.